sklearn/impute.py

```diff
@@ -873,3 +874,201 @@ def fit(self, X, y=None):
+                             "than during fitting.")
+
+        imputer_mask, features_ = self._get_missing_features_info(X)
+
+        if isinstance(self.features, six.string_types):
+            if self.features == "auto":
+                features = np.setdiff1d(features_,
+                                        self.features_)
+                if self.error_on_new and features.size > 0:
+                    raise ValueError("The features %s have missing values "
+                                     "in transform but have no missing values "
+                                     "in fit" % features)
+
+        if not (isinstance(self.features, six.string_types) and
+                self.features == "all") and len(self.features_) != 0:
+            # no need to slice when all features have missing values
+            imputer_mask = imputer_mask[:, self.features_]
+
+        return imputer_mask
+
+    def fit_transform(self, X, y=None):
+        """Generate missing values indicator for X.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = [n_samples, n_features]
+            The input data to complete.
+
+        Returns
+        -------
+        Xt : array or sparse matrix, shape = [n_samples, n_features]
+             The missing indicator for input data
+
+        """
+        self.fit(X, y)
+        return self.transform(X)
+
+    def _validate_params(self):
+        if (isinstance(self.features, six.string_types) and
+                self.features not in ["auto", "all"]):
+            raise ValueError("Can only use these options: 'auto', 'all'"
+                             " got {0}".format(self.features))
+        elif not isinstance(self.features, six.string_types):
+            self.features = check_array(self.features, ensure_2d=False)
+            if (isinstance(self.features, np.ndarray) and
+                    self.features.dtype.kind != 'i'):
+                raise ValueError("Features should be an array of integers")
+
+        if not ((isinstance(self.sparse, six.string_types) and
+                self.sparse == "auto") or isinstance(self.sparse, bool)):
+            raise ValueError("sparse can only boolean or 'auto'"
+                             " got {0}".format(self.sparse))
+
+    def _get_missing_features_info(self, X):
+        if sparse.issparse(X) and self.missing_values != 0:
+            #  sparse matrix and missing values is not zero
+            imputer_mask = _get_mask(X.data, self.missing_values)
+            imputer_mask = X.__class__((imputer_mask, X.indices.copy(),
+                                        X.indptr.copy()), shape=X.shape,
+                                       dtype=X.dtype)
+            feat_with_missing = imputer_mask.sum(axis=0).nonzero()[1]
+        else:
+            #  sparse with zero as missing value and dense matrix
+            if sparse.issparse(X):
+                X = X.toarray()
```

Comment:
@jnothman I am not sure if we want to convert it. It depends if we consider the explicit 0 as missing value.
For instance it would be more efficient to do something like:

```python
                features_without_missing = (np.flatnonzero(np.diff(X.indptr))        
                                            if X.format == 'csc'                     
                                            else np.unique(X.indices))               
                feat_with_missing = np.setdiff1d(np.arange(X.shape[1]),              
                                                 features_without_missing)           
                imputer_mask = (~X.toarray().astype(dtype=bool))                     
                imputer_mask = imputer_mask.astype(int, copy=False)
```